src/http/http_client.cpp

```diff
@@ -0,0 +1,336 @@
+}
+
+dsn::error_s http_client::with_get_method() { return set_method(http_method::GET); }
+
+dsn::error_s http_client::set_method(http_method method)
+{
+    // No need to process the case of http_method::POST, since it should be enabled by
+    // `with_post_method`.
+    switch (method) {
+    case http_method::GET:
+        RETURN_IF_SETOPT_NOT_OK(CURLOPT_HTTPGET, 1L);
+        break;
+    default:
+        LOG_FATAL("Unsupported http_method");
+    }
+
+    _method = method;
+    return dsn::error_s::ok();
+}
+
+dsn::error_s http_client::set_timeout(long timeout_ms)
+{
+    RETURN_IF_SETOPT_NOT_OK(CURLOPT_TIMEOUT_MS, timeout_ms);
+    return dsn::error_s::ok();
+}
+
+void http_client::clear_header_fields()
+{
+    _header_fields.clear();
+
+    _header_changed = true;
+}
+
+void http_client::free_header_list()
+{
+    if (_header_list == nullptr) {
+        return;
+    }
+
+    curl_slist_free_all(_header_list);
+    _header_list = nullptr;
+}
+
+void http_client::set_header_field(dsn::string_view key, dsn::string_view val)
+{
+    _header_fields[std::string(key)] = std::string(val);
+    _header_changed = true;
+}
+
+void http_client::set_accept(dsn::string_view val) { set_header_field("Accept", val); }
+
+void http_client::set_content_type(dsn::string_view val) { set_header_field("Content-Type", val); }
+
+dsn::error_s http_client::process_header()
+{
+    if (!_header_changed) {
+        return dsn::error_s::ok();
+    }
+
+    free_header_list();
+
+    for (const auto &field : _header_fields) {
+        auto str = fmt::format("{}: {}", field.first, field.second);
+
+        // A null pointer is returned if anything went wrong, otherwise the new list pointer is
+        // returned. To avoid overwriting an existing non-empty list on failure, the new list
+        // should be returned to a temporary variable which can be tested for NULL before updating
+        // the original list pointer. (https://curl.se/libcurl/c/curl_slist_append.html)
+        struct curl_slist *temp = curl_slist_append(_header_list, str.c_str());
+        if (temp == nullptr) {
+            free_header_list();
+            return dsn::error_s::make(dsn::ERR_CURL_FAILED, "curl_slist_append failed");
+        }
+        _header_list = temp;
+    }
+
+    // This would work well even if `_header_list` is NULL pointer. Pass a NULL to this option
+    // to reset back to no custom headers. (https://curl.se/libcurl/c/CURLOPT_HTTPHEADER.html)
+    RETURN_IF_SETOPT_NOT_OK(CURLOPT_HTTPHEADER, _header_list);
+
+    // New header has been built successfully, thus mark it unchanged.
+    _header_changed = false;
+
+    return dsn::error_s::ok();
+}
+
+dsn::error_s http_client::exec_method(const http_client::recv_callback &callback)
+{
+    // `curl_easy_perform` would run synchronously, thus it is safe to use the pointer to
+    // `callback`.
+    _recv_callback = &callback;
+
+    const auto &err = process_header();
+    if (!err) {
+        return err;
+    }
```

Comment:
How about use macro `RETURN_NOT_OK` to simplify the code?

Reply:
OK.